pkg/migration/migrations/fix_privileges_migration.go

```diff
@@ -0,0 +1,172 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/errors"
+)
+
+func fixPrivilegesMigration(
+	ctx context.Context, _ clusterversion.ClusterVersion, d migration.TenantDeps,
+) error {
+	var lastUpgradedID descpb.ID
+	for {
+		done, idToUpgrade, objectType, err := findNextDescriptorToFix(ctx, d.InternalExecutor, lastUpgradedID)
+		if err != nil || done {
+			return err
+		}
+		if err := fixPrivileges(ctx, idToUpgrade, d, objectType); err != nil {
+			return err
+		}
+		lastUpgradedID = idToUpgrade
+	}
+}
+
+func fixPrivileges(
+	ctx context.Context, upgrade descpb.ID, d migration.TenantDeps, objectType privilege.ObjectType,
+) error {
+	return descs.Txn(ctx, d.Settings, d.LeaseManager, d.InternalExecutor, d.DB, func(
+		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
+	) error {
+		var desc catalog.MutableDescriptor
+		var err error
+		var privilegeDesc *descpb.PrivilegeDescriptor
+		switch objectType {
+		case privilege.Table:
+			desc, err = descriptors.GetMutableTableByID(ctx, txn, upgrade, tree.ObjectLookupFlagsWithRequired())
+			if err != nil {
+				return err
+			}
+			privilegeDesc = desc.GetPrivileges()
+		case privilege.Database:
+			desc, err = descriptors.GetMutableDatabaseByID(ctx, txn, upgrade, tree.DatabaseLookupFlags{Required: true})
+			if err != nil {
+				return err
+			}
+			privilegeDesc = desc.GetPrivileges()
+		case privilege.Schema:
+			desc, err = descriptors.GetMutableSchemaByID(ctx, txn, upgrade, tree.SchemaLookupFlags{Required: true})
+			if err != nil {
+				return err
+			}
+			privilegeDesc = desc.GetPrivileges()
+		case privilege.Type:
+			desc, err = descriptors.GetMutableTypeByID(ctx, txn, upgrade, tree.ObjectLookupFlags{
+				CommonLookupFlags: tree.CommonLookupFlags{Required: true}},
+			)
+			privilegeDesc = desc.GetPrivileges()
+		}
+
+		if privilegeDesc.Version >= descpb.Version21_2 {
+			log.Infof(ctx, "privilege descriptor has already been fixed, skipping %d", upgrade)
+			return nil
+		}
+
+		for i := range privilegeDesc.Users {
+			privilegeDesc.Users[i].Privileges &= privilege.GetValidPrivilegesForObject(objectType).ToBitField()
+		}
+		if privilegeDesc.Owner().Undefined() {
+			if desc.GetID() == keys.SystemDatabaseID || desc.GetParentID() == keys.SystemDatabaseID {
+				privilegeDesc.SetOwner(security.NodeUserName())
+			}
+			privilegeDesc.SetOwner(security.RootUserName())
+		}
+
+		if privilegeDesc.Version < descpb.Version21_2 {
+			privilegeDesc.SetVersion(descpb.Version21_2)
+		}
+
+		return descriptors.WriteDesc(ctx, false /* kvTrace */, desc, txn)
```

Comment:
I allude to this in another comment but wouldn't having one transaction per descriptor be wasteful?